### ISSUE TITLE
Unordered observations might call 'added' before 'removed' on the client

### ISSUE THREAD

**Issue by TeNNoX** [meteor/meteor#10916]
I have **a server method that fully deletes a collection with >1000 documents and re-inserts them** (with different MongoIDs, and some different fields). And I have observers that keep another (local) collection in sync.
On the server, the observers are called in expected order, but on the client, I'm sometimes **observing 'added' events before their respective 'removed' events**.

I spent quite some hours of debugging, and I at least got a theory of why it happens, although I failed to proof it until now. Here's what I observed:
- in `livedata_connection`, the `_bufferedWrites` are in correct order.
- the `local_collection`'s `query` is `ordered: false`
- (I think) the collection's `_observeQueue`'s tasks are in wrong order
- (I'm sure) the observe callbacks are called `added` before the respective `removed`

And another important observation: **when I force the query to be ordered** - I added an empty `changedAt` observer to the mix - **all observations are called in the correct order**!

This, so I think, is caused by `DiffSequence.diffQueryUnorderedChanges` (called by `local_collection.resumeObservers`), which [executes all `added` callbacks before calling all `removed`](https://github.com/meteor/meteor/blob/31923642171aabb1bdc95b30c1f0a5ee16af4c6e/packages/diff-sequence/diff.js#L37)

**Meteor**: 1.9.3
**OS**: elementary OS 5.1 Hera

**Comment by TeNNoX** [comment]
I also managed to reproduce the problem when using ordered observers:

```javascript
// merge logs of same type for greater clarity
export const logs = [];
function log(t, m) {
  console.log(t, m);
  if (logs.length && logs[logs.length - 1].t === t) logs[logs.length - 1].m += `, ${m}`;
  else logs.push({ t, m });
}

export const testMethod = new ValidatedMethod({
  name: 'testMethod',
  validate: new SimpleSchema({}).validator(),
  run() {
    const ids = [];
    for (let i = 0; i < 500; i++) {
      ids.push(
        Elements.insert({
          name: `${i}`,
        })
      );
    }
    if (Meteor.isClient) log('TEST', 'simulation done');
    if (Meteor.isServer) {
      for (let i = Meteor.isServer ? 100 : 200; i < 400; i++) {
        Elements.remove({
          _id: ids[i],
        });
      }
      for (let i = 100; i < 150; i++) {
        Elements.insert({
          name: `${i}`,
        });
      }
    }
  },
});

if (Meteor.isClient) {
  const C = new Set();
  Elements.find().observe({
    addedAt: model => {
      log('ADD', `${C.has(model.name) ? 'X' : ''}${model.name}`);
      LV >= 7 && console.debug('ADD', model);
      C.add(model.name);
    },
    removedAt: model => {
      const had = C.delete(model.name) ? '' : 'X';
      log('REM', `${had}${model.name}`);
      LV >= 7 && console.debug('REM', model);
    },
  });
}
```

The result is (simplified):

> ADD 0, 1, [...] 497, 498, 499
> TEST simulation done
> ADD X100, X101, X102, [...] X124, X125, X126
> REM 100, 101, [...] 398, 399
> ADD 127, 128, [...] 148, 149

As you can see, some of the **ADD** events are fired before the **REMOVED** events - causing a conflict as marked by the **`X`**.

**Comment by TeNNoX** [comment]
I've found another hint while examining the DDP messages:

> added: xfRqNCoFtS7ixcXdL {"name":"0"}
> ...
> added: RT8sNteCWS7FcWf8H {"name":"498"}
> added: Z4QLS2Dirwtykkpih {"name":"499"}
> removed: 6zLNbbthDoSGHsnrh
> removed: mcrabv3gRtzjbMAnD
> ...
> removed: wR6mCBT7ghF4LGprb
> added: Yyfo29HBBbxjwDwfq {"name":"100"}
> added: N8f6XnNrEZeRLYvCZ {"name":"101"}
> ...
> added: wvbiGvt5u5ZC7DNrN {"name":"126"}
> result: {"id":"2","result":"done"}
> added: re4SquEE2aycjBuWR {"name":"127"}
> ...
> added: CkysCi5YMpRqr5tz2 {"name":"148"}
> added: SFxYoajQC878mfyaS {"name":"149"}
> updated

As you can see the batch of **ADDED** events that come before the **REMOVED** are also the ones that come before the DDP **`result`** message

**Comment by sebakerckhof** [comment]
I'll try to take a look this week.

**Comment by sebakerckhof** [comment]
Just took a look and the behavior is correct.

Meteor guarantees you will never see a removed for a certain mongo object, before you see an added. However, in this case, these are different objects (just with the same label, but different _id) and Meteor makes no guarantees around those.

This has to do with performance optimizations on both frontend and backend:
When a large amount of changes come in over DDP, DDP will update these on the local collection as a batch. When minimongo gets a batch update, it will pause its observers. After the batch is completed it will do a diff to find what has changed and call the relevant callbacks. It does this to prevent heavy load & flickering in the presentation layer (which would be true for most UI frameworks today). However, order of adds & removed is lost in this process (except for objects which are the same, based on _id).

Above explains the front-end. You may now think you can just tap into the DDP messages which do arrive in order. But even this should not be trusted because of another performance optimization Meteor does on the backend. Namely, when the oplog grows too fast for Meteor to process, it will stop doing oplog tailing and fall back to poll & diff on the backend. Meaning it will just run the query every X seconds and diff the results in the same way the frontend does it when it gets a batch update. And queries which cannot use oplog tailing would always have this behavior anyway.

So in short, the order of adds/removed should not be relied on as Meteor makes no guarantees about those because of the processes described above. Meteor does guarantee you will not see a removed or changed (or an older change after a newer change) before an added for the same object (= same _id).

**Comment by TeNNoX** [comment]
Wow! Thank you for this extensive and comprehensive answer :+1: 

I see, so I need to adapt my client-side code in accordance to this.
Is there a document with those guarantees and details? That would be amazing, and ease development a lot!

For example I had the question if there are any kind of guarantees on ordering of events for documents that are not the same `_id`?
From how I understand you, the only guarantee is there will be no illegal ordering (e.g. removing when not existent) for a single document's changes. Is that all?

Thanks for your work, all in all - it helps a ton!

**Comment by gotjoshua** [comment]
> And queries which cannot use oplog tailing 
Is addition to documentation that @TeNNoX asked for, I'm rather curious about which queries can and can't use oplog tailing. Could you provide links to code and/or documentation about that?

**Comment by sebakerckhof** [comment]
yeah, I've been planning to do a blog post or video to demystify some parts of the livequery system. So people can make better decisions regarding performance etc. But haven't found the time for it.

Regarding the question of @TeNNoX : Yes, to the best of my knowledge, the only guarantee is on the ordering of events on the same document.

Regarding the question of @gotjoshua . This piece of code determines whether oplog tailing is used:
https://github.com/meteor/meteor/blob/devel/packages/mongo/mongo_driver.js#L1289-L1323
But it doesn't tell you very much. It depends on the query options and what minimongo supports, which isn't documented comprehensively. This is a non-comprehensive list:
https://galaxy-guide.meteor.com/apm-optimize-your-app-for-oplog.html#When-You-can%E2%80%99t-Use-Oplog
